### ISSUE TITLE
Search for posts not finding posts with specific keywords

### ISSUE THREAD

**Issue by ori-near** [near/neardevhub-widgets#106]
**❌ Dependency**
This is dependent on https://github.com/near/neardevhub-widgets/issues/115 getting completed first.

**Problem**
Although we recently introduced search capability in #66, the current design implementation has some [limitations](https://github.com/near/neardevhub-widgets/pull/85#issuecomment-1500202212) that lead to a poor user experience. To name a few examples:

- The search only works for top level parent posts, meaning that users are unable to search for child level posts
- The fuzzy search sometimes omits explicit terms
- The search experience is slow and results in freezes such as when searching for long text queries

**User Story**
As a DevHub user, I want to search for any posts that mention a specific topic or keyword even if I don't know the exact phrasing or spelling.

**Acceptance Criteria**

- The search works for any posts (top level or child)
- The search function uses fuzzy matching to return posts that match the search term, including partial matches and misspelling. The search should always prioritize the explicit terms first
- The search results should load quickly
- The search function displays results as they are found, allowing users to see posts matching the search in real-time

---

### Example of searching for terms in child posts

**To reproduce:**
Here's an example of child terms that get omitted:

1. Log into the [DevHub](https://near.social/#/devgovgigs.near/widget/gigs-board.pages.Feed) as theori.near
2. Filter by `zero-knowledge` labels and the `devgovgigs.near` author
3. Search for "typically"
4. Search output returns: "No posts matches search"

**Expected behavior:**
Search output should reveal 2 posts that match these conditions ([post 1](https://near.social/#/devgovgigs.near/widget/gigs-board.pages.Post?id=372), [post 2](https://near.social/#/devgovgigs.near/widget/gigs-board.pages.Post?id=373))

**Desktop:**
OS: OS
Browser Chrome
Version: 112.0.5615.49

**Screenshots:**
![Screenshot 2023-04-20 at 6 10 34 PM](https://user-images.githubusercontent.com/110252255/233517531-1774cbe6-f825-4685-8a2b-c94a9d1e897b.png)

https://user-images.githubusercontent.com/110252255/233517605-6e96a58f-371f-4a32-a223-e5378aab7523.mp4

**Comment by frol** [comment]
@ori-near Just for context, both of the posts are not top-level. The solution we currently have requires downloading all the posts to the client side, so it was decided to only use top-level posts for search, which I highlighted as a limitation in the PR review https://github.com/near/devgigsboard-widgets/pull/85#issuecomment-1500202212. I will leave it to @ailisp to evaluate the best path forward, but I feel that soon we will need a dedicated server to serve search queries.

**Comment by ailisp** [comment]
@frol 's observation is correct. Searching for typically finds top level posts, which is expected behavior. If you find some keyword where top level posts are not found in search result, please feel free to reopen this ticket!

**Comment by frol** [comment]
I would still keep this issue open, so we can get back to it in the future

**Comment by ailisp** [comment]
Done. Note that server side processing is required to complete this task. So I create a separate task near/neardevhub-contract#21, which this task is consider blocked by.
